File: 3.Ottobre/72.ThermostatAdjkuster2-21Ottobre.js
//todo  Data la temperatura attuale di una stanza in gradi Fahrenheit e una temperatura target in gradi Celsius, restituisci una stringa che indica come regolare la temperatura ambiente in base a questi vincoli:

//! Restituisci "Riscaldamento: X gradi Fahrenheit" se la temperatura attuale è inferiore al target. Dove X è il numero di gradi Fahrenheit necessari per riscaldare la stanza e raggiungere il target, arrotondato a 1 cifra decimale.
//! Restituisci "Raffreddamento: X gradi Fahrenheit" se la temperatura attuale è superiore al target. Dove X è il numero di gradi Fahrenheit necessari per raffreddare la stanza e raggiungere il target, arrotondato a 1 cifra decimale.
//! Restituisci "Mantieni" se la temperatura attuale è uguale al target.
//! Per convertire Celsius in Fahrenheit, moltiplica la temperatura Celsius per 1,8 e aggiungi 32 al risultato (F = (C * 1,8) + 32).

//? Tempo di esecuzione:
//* 15 Minuti e 36 secondi 

//?Codice scritto senza utilizzo di ChatGpt

function adjustThermostat(currentF, targetC) {

  let celsiusConvert = (targetC * 1.8) + 32
  let difference = celsiusConvert - currentF

  if(difference === 0) return  "Hold"
  if(difference > 0)   return  "Heat:" + " " + difference.toFixed(1) + " " + "degrees Fahrenheit"
  if(difference < 0)   return  "Cool:" + " " + Math.abs(difference.toFixed(1)) + " " + "degrees Fahrenheit"
}

console.log(adjustThermostat(72, 18))

//?Codice scritto da ChatGpt
//* Non usata

//* ✅ PRO

//* Conversione C → F corretta

//* Logica difference chiara

//* Gestione dei 3 casi (Heat / Cool / Hold)

//* toFixed(1) usato correttamente

//* Output conforme alla traccia

//! ❌ CONTRO

//! Math.abs(difference.toFixed(1)) → toFixed restituisce stringa
//! (meglio Math.abs(difference).toFixed(1))

//! Stringhe concatenate inutilmente ("Heat:" + " " + ...) 

//! Mancano spazi gestiti in modo più pulito (template literal)

//! Nessun else finale (non grave, ma migliorabile)

//? Versione Chat Gpt 

function adjustThermostat(currentF, targetC) {
  const targetF = targetC * 1.8 + 32
  const diff = targetF - currentF

  if (diff === 0) return "Hold"
  if (diff > 0) return `Heat: ${diff.toFixed(1)} degrees Fahrenheit`
  return `Cool: ${Math.abs(diff).toFixed(1)} degrees Fahrenheit`
}

//todo Considerazioni: 

//! N.B.

File: Python/3.Ottobre/72.Thermostat_adjkuster2-21Ottobre.py
# =========================================================
# FILE TEMPLATE - ANALISI ESERCIZI PYTHON
# =========================================================

# =========================================================
# //todo Consegna dell'esercizio
# =========================================================

# Data la temperatura attuale in Fahrenheit e una temperatura
# target in Celsius, restituire:
#
# - "Heat: X degrees Fahrenheit" se bisogna riscaldare.
# - "Cool: X degrees Fahrenheit" se bisogna raffreddare.
# - "Hold" se la temperatura è già quella desiderata.
#
# Convertire prima i gradi Celsius in Fahrenheit con:
# F = (C * 1.8) + 32
#
# X deve essere arrotondato a una cifra decimale.

# =========================================================
# //* Codice scritto senza utilizzo di ChatGpt
# =========================================================

def adjust_thermostat(current_f, target_c):
    to_Fahrenheit = (target_c * 1.8) + 32

    if current_f == to_Fahrenheit:
        return "Hold"
    elif current_f < to_Fahrenheit:
        return f"Heat: {round(to_Fahrenheit - current_f, 1)} degrees Fahrenheit"
    elif current_f > to_Fahrenheit:
        return f"Cool: {round(current_f - to_Fahrenheit, 1)} degrees Fahrenheit"

print(adjust_thermostat(72, 18))

# =========================================================
# //* Codice scritto da ChatGpt (versione migliorata)
# =========================================================

def adjust_thermostat(current_f, target_c):
    target_f = target_c * 1.8 + 32
    difference = round(abs(target_f - current_f), 1)

    if current_f == target_f:
        return "Hold"
    if current_f < target_f:
        return f"Heat: {difference} degrees Fahrenheit"
    return f"Cool: {difference} degrees Fahrenheit"

# =========================================================
# //* Pro e contro del mio codice
# =========================================================

# //* Pro
# - Conversione da Celsius a Fahrenheit corretta.
# - Logica dei confronti chiara e ben organizzata.
# - Arrotondamento a una cifra decimale corretto.
# - Utilizzo delle f-string appropriato.

# //! Contro
# - Il nome della variabile to_Fahrenheit può essere semplificato in target_f.
# - Il calcolo della differenza viene ripetuto in due rami dell'if.
# - Dopo un return non è necessario usare elif: basta un nuovo if.
# - I ";" finali non servono in Python.

# =========================================================
# //* Consigli
# =========================================================

# - Evita i ";" a fine riga: sono ereditati da altri linguaggi ma in Python sono inutili.
# - Quando lo stesso calcolo compare più volte, salvalo in una variabile.
# - Usa nomi brevi ma descrittivi (target_f è più leggibile di to_Fahrenheit).
# - Ricorda che dopo un return la funzione termina, quindi spesso gli elif possono essere sostituiti da semplici if.